### DISCUSSION TITLE
Question: Modal with Interactivity API

### DISCUSSION THREAD

**Discussion by TheAggressive** [WordPress/gutenberg#63907] [Interactivity API]
Looking to build a modal block with the interactivity api but want to know what’s the best way to approach it? How would you guys structure it?

As I see it, to make a modal I have a couple options. Create (2) blocks. One block would controls (items that can be clicked to change state, etc.) and the other block to listen for changes in the state of the block above and structures the modal that will be placed in the footer. 

Or create one block. Then inject the modal markup with directives and hook it to the wp_footer (with hooks) and show based off of the state. If that makes sense?

Or do you guys have any other ideas how this could be implemented with the IAPI?

Thanks for any feedback!

**Comment by luisherranz** [comment]
It's about what works best for you, but it's true that for the `Expand on click`, feature of the Image block, we added the modal to the `body` using the `wp_footer` action, and that approach has worked well.

- Action in the render callback: https://github.com/WordPress/gutenberg/blob/f28496737d503fb5395ba8543922596075485e4d/packages/block-library/src/image/index.php#L261
- Modal: https://github.com/WordPress/gutenberg/blob/f28496737d503fb5395ba8543922596075485e4d/packages/block-library/src/image/index.php#L269-L324

**Comment by djcowan** [comment]
I'm currently looking at developing a modal block and am keen to implement using the HTMLDialogElement `showModal` method rather than `data-wp-bind--open`... at least on the frontend. `showModal` exposes the ::background pseudo element which can be styled and animated using vanilla css. The editor experience will require a different approach as it can be difficult to add innerblocks, or access the block controls when the modal is obscuring the editor. Discovered this with my initial attempt to utilise the existing modal component. Next step is to find a way to bind the dialog to the trigger and probably learn javascript.

**Reply by TheAggressive** [reply]
Would be cool to see how you implement this! I was considering this approach too.

**Reply by djcowan** [reply]
at the moment.. literally a moment ago; I has some success using a `data-wp-watch` callback. However, I don't know if this is the correct path.
___
// file:./render.php
```php
<aside  data-wp-interactive="acme-dialog" <?php echo \wp_interactivity_data_wp_context(['isOpen' => false]); ?> >
  <dialog data-wp-watch="callbacks.openModal" <?php echo \get_block_wrapper_attributes(); ?> >
     <form method="dialog">
       <button type="submit" data-wp-on-async--click="actions.toggle">close</button>
      </form>
        <div>
            <?php echo $content; ?><!-- innerblocks -->
        </div>
  </dialog>
  <button data-wp-on-async--click="actions.toggle">open</button>
</aside>
```
___

// file:./view.js

```tsx
import { store, getContext, getElement } from '@wordpress/interactivity';

const { state } = store( "acme-dialog", {
	state: {
		get isOpen()
		{
			const context: { isOpen: boolean; } = getContext();
			return context.isOpen;
		},
	},
	actions: {
		toggle: ( event: React.ChangeEvent<HTMLInputElement> ) =>
		{
			const context: any = getContext();
			context.isOpen = !context.isOpen;
		}
	},
	callbacks: {
		openModal: () =>
		{
			// const { isOpen, targetId }: { isOpen: boolean; targetId: string; } = getContext();
			// const { ref }: { ref: HTMLDialogElement; } = getElement(); // @todo #12 typescript incompat issue ref: HTMLElement type
			const { ref }: { ref: any; } = getElement();
			if ( state.isOpen )
			{
				ref?.showModal();
			}
		},
	},

} );
```
___
the callback `openModal` is watching for a change to the state of `isOpen` and triggers `showModal()` on change. This is barebones and will require probably require some additional conditional checks. But like I said; javascript - for me - is a work in progress.
Would love to here thoughts and feedback on the approach taken. 
![Untitled](https://github.com/user-attachments/assets/d5952462-5b57-41ba-9844-ace9412a7305)

**Reply by djcowan** [reply]
oh..the edit function uses the HTMLelement `dialog` but the `open` state is not triggered via js - which render the block without a backdrop, allowing innerblocks to be inserted etc.
```jsx
// fn Edit() {
	const [ isOpen, setIsOpen ] = useState( false );

 return ( // ...
			{ isOpen && (
				<dialog { ...blockProps } open={ isOpen }>
					<form method="dialog">
						<Flex className={ `${ className }__header` }>
						// ...
				<Button onClick={ (): void => setIsOpen( !isOpen ) }
// ...
)
```

**Reply by luisherranz** [reply]
> I has some success using a `data-wp-watch` callback. However, I don't know if this is the correct path.

It is correct since it is a declarative way to subscribe to a state, linking it to a modification of the DOM.

You can also close it with the same callback.

```js
const { state } = store('acme-dialog', {
  callbacks: {
    controlModal() {
      const { ref } = getElement();
      if (state.isOpen) {
        ref?.showModal();
      } else {
        ref?.close();
      }
    },
  },
});
```

**Reply by djcowan** [reply]
Thank you. Appreciate you taking the time to respond and assuage my doubt. I must add the interactivity API is well thought out and surprisingly simple to interact with.

**Reply by luisherranz** [reply]
Thanks 🙂

**Comment by helgatheviking** [comment]
@luisherranz A little off-topic, but if I were going to write a modal, how viable do you think it is to re-use parts of the image block iAPI? Or would it be best to duplicate everything? I am specifically think about the opening/closing/focus/styling of the overlay itself and not necessarily the content.

**Reply by luisherranz** [reply]
For now, I would duplicate everything. In the future, if we see that this is a specific pattern, we can provide an abstraction to make it very easy to create accessible modals with the Interactivity API.

**Reply by helgatheviking** [reply]
Thank you Luis!
